src/app/codegen-data-model-provider/CodegenDataModelProvider.cpp

```diff
@@ -433,51 +482,58 @@ std::optional<unsigned> CodegenDataModelProvider::TryFindEndpointIndex(EndpointI
     return std::make_optional<unsigned>(idx);
 }
 
-EndpointId CodegenDataModelProvider::NextEndpoint(EndpointId before)
+DataModel::EndpointEntry CodegenDataModelProvider::NextEndpoint(EndpointId before)
 {
     const uint16_t lastEndpointIndex = emberAfEndpointCount();
 
     std::optional<unsigned> before_idx = TryFindEndpointIndex(before);
     if (!before_idx.has_value())
     {
-        return kInvalidEndpointId;
+        return DataModel::EndpointEntry::kInvalid;
     }
 
     // find the first enabled index
     for (uint16_t endpoint_idx = static_cast<uint16_t>(*before_idx + 1); endpoint_idx < lastEndpointIndex; endpoint_idx++)
     {
         if (emberAfEndpointIndexIsEnabled(endpoint_idx))
         {
-            mEndpointIterationHint = endpoint_idx;
-            return emberAfEndpointFromIndex(endpoint_idx);
+            mEndpointIterationHint                 = endpoint_idx;
+            DataModel::EndpointEntry endpointEntry = DataModel::EndpointEntry::kInvalid;
+            endpointEntry.id                       = emberAfEndpointFromIndex(endpoint_idx);
+            auto endpointInfo                      = GetEndpointInfoAtIndex(endpoint_idx);
+            VerifyOrDie(endpointInfo.has_value());
+            endpointEntry.info = endpointInfo.value();
+            return endpointEntry;
         }
     }
 
     // No enabled enpoint after "before" was found, give up
-    return kInvalidEndpointId;
+    return DataModel::EndpointEntry::kInvalid;
 }
 
-DataModel::ClusterEntry CodegenDataModelProvider::FirstCluster(EndpointId endpointId)
+DataModel::ClusterEntry CodegenDataModelProvider::FirstServerCluster(EndpointId endpointId)
 {
     const EmberAfEndpointType * endpoint = emberAfFindEndpointType(endpointId);
     VerifyOrReturnValue(endpoint != nullptr, DataModel::ClusterEntry::kInvalid);
     VerifyOrReturnValue(endpoint->clusterCount > 0, DataModel::ClusterEntry::kInvalid);
     VerifyOrReturnValue(endpoint->cluster != nullptr, DataModel::ClusterEntry::kInvalid);
 
-    return FirstServerClusterEntry(endpointId, endpoint, 0, mClusterIterationHint);
+    return FirstServerClusterEntry(endpointId, endpoint, 0, mServerClusterIterationHint);
 }
 
-std::optional<unsigned> CodegenDataModelProvider::TryFindServerClusterIndex(const EmberAfEndpointType * endpoint,
-                                                                            ClusterId id) const
+std::optional<unsigned> CodegenDataModelProvider::TryFindClusterIndex(const EmberAfEndpointType * endpoint, ClusterId id,
+                                                                      bool isServer) const
```

Comment:
please make a `ClusterSide` or `ClusterType` enum class and pass that, so we compare to `kClientCluster` and `kServerCluster` instead of just a `bool`. Bools in a list of arguments is often unclear. I.e::

```cpp
TryFindClusterIndex(endpointType, clusterId, true);
```

is much harder to immediately read than:

```
TryFindClusterIndex(endpointType, clusterId, kServerClusterType);
```

Reply:
Done